### ISSUE TITLE
storage :file fail with QEMU session

### ISSUE THREAD

**Issue by dkrm0** [vagrant-libvirt/vagrant-libvirt#986]
### Steps to reproduce
1. Use QEMU session with `libvirt.qemu_use_session = true`
2. Create and attach an additional disks via `libvirt.storage :file`
3. Vagrant will crash, trying to create the domain

### Expected behaviour
The domain should be created with an additionnal disk.

### Actual behaviour
Vagrant crash with this backtrace:
```
/home/dkrm/vms/vagrant/gems/2.6.2/gems/vagrant-libvirt-0.0.45/lib/vagrant-libvirt/action/create_domain.rb:156:in `block in call': undefined method `+' for nil:NilClass (NoMethodError)
	from /home/dkrm/vms/vagrant/gems/2.6.2/gems/vagrant-libvirt-0.0.45/lib/vagrant-libvirt/action/create_domain.rb:147:in `each'
	from /home/dkrm/vms/vagrant/gems/2.6.2/gems/vagrant-libvirt-0.0.45/lib/vagrant-libvirt/action/create_domain.rb:147:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-2.2.4/lib/vagrant/action/warden.rb:50:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-2.2.4/lib/vagrant/action/builtin/before_trigger.rb:23:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-2.2.4/lib/vagrant/action/warden.rb:50:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-2.2.4/lib/vagrant/action/builtin/after_trigger.rb:26:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-2.2.4/lib/vagrant/action/warden.rb:50:in `call'
	from /home/dkrm/vms/vagrant/gems/2.6.2/gems/vagrant-libvirt-0.0.45/lib/vagrant-libvirt/action/create_domain_volume.rb:83:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-2.2.4/lib/vagrant/action/warden.rb:50:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-2.2.4/lib/vagrant/action/builtin/before_trigger.rb:23:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-2.2.4/lib/vagrant/action/warden.rb:50:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-2.2.4/lib/vagrant/action/builtin/after_trigger.rb:26:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-2.2.4/lib/vagrant/action/warden.rb:50:in `call'
	from /home/dkrm/vms/vagrant/gems/2.6.2/gems/vagrant-libvirt-0.0.45/lib/vagrant-libvirt/action/handle_box_image.rb:146:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-2.2.4/lib/vagrant/action/warden.rb:50:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-2.2.4/lib/vagrant/action/builtin/before_trigger.rb:23:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-2.2.4/lib/vagrant/action/warden.rb:50:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-2.2.4/lib/vagrant/action/builtin/after_trigger.rb:26:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-2.2.4/lib/vagrant/action/warden.rb:50:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-2.2.4/lib/vagrant/action/builtin/handle_box.rb:56:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-2.2.4/lib/vagrant/action/warden.rb:50:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-2.2.4/lib/vagrant/action/builtin/before_trigger.rb:23:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-2.2.4/lib/vagrant/action/warden.rb:50:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-2.2.4/lib/vagrant/action/builtin/after_trigger.rb:26:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-2.2.4/lib/vagrant/action/warden.rb:50:in `call'
	from /home/dkrm/vms/vagrant/gems/2.6.2/gems/vagrant-libvirt-0.0.45/lib/vagrant-libvirt/action/handle_storage_pool.rb:57:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-2.2.4/lib/vagrant/action/warden.rb:50:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-2.2.4/lib/vagrant/action/builtin/before_trigger.rb:23:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-2.2.4/lib/vagrant/action/warden.rb:50:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-2.2.4/lib/vagrant/action/builtin/after_trigger.rb:26:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-2.2.4/lib/vagrant/action/warden.rb:50:in `call'
	from /home/dkrm/vms/vagrant/gems/2.6.2/gems/vagrant-libvirt-0.0.45/lib/vagrant-libvirt/action/set_name_of_domain.rb:35:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-2.2.4/lib/vagrant/action/warden.rb:50:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-2.2.4/lib/vagrant/action/builtin/before_trigger.rb:23:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-2.2.4/lib/vagrant/action/warden.rb:50:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-2.2.4/lib/vagrant/action/warden.rb:121:in `block in finalize_action'
	from /opt/vagrant/embedded/gems/gems/vagrant-2.2.4/lib/vagrant/action/warden.rb:50:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-2.2.4/lib/vagrant/action/builder.rb:116:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-2.2.4/lib/vagrant/action/runner.rb:102:in `block in run'
	from /opt/vagrant/embedded/gems/gems/vagrant-2.2.4/lib/vagrant/util/busy.rb:19:in `busy'
	from /opt/vagrant/embedded/gems/gems/vagrant-2.2.4/lib/vagrant/action/runner.rb:102:in `run'
	from /opt/vagrant/embedded/gems/gems/vagrant-2.2.4/lib/vagrant/action/builtin/call.rb:53:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-2.2.4/lib/vagrant/action/warden.rb:50:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-2.2.4/lib/vagrant/action/builtin/before_trigger.rb:23:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-2.2.4/lib/vagrant/action/warden.rb:50:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-2.2.4/lib/vagrant/action/builtin/after_trigger.rb:26:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-2.2.4/lib/vagrant/action/warden.rb:50:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-2.2.4/lib/vagrant/action/builtin/box_check_outdated.rb:84:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-2.2.4/lib/vagrant/action/warden.rb:50:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-2.2.4/lib/vagrant/action/builtin/before_trigger.rb:23:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-2.2.4/lib/vagrant/action/warden.rb:50:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-2.2.4/lib/vagrant/action/builtin/after_trigger.rb:26:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-2.2.4/lib/vagrant/action/warden.rb:50:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-2.2.4/lib/vagrant/action/builtin/config_validate.rb:25:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-2.2.4/lib/vagrant/action/warden.rb:50:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-2.2.4/lib/vagrant/action/builtin/before_trigger.rb:23:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-2.2.4/lib/vagrant/action/warden.rb:50:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-2.2.4/lib/vagrant/action/builder.rb:116:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-2.2.4/lib/vagrant/action/runner.rb:102:in `block in run'
	from /opt/vagrant/embedded/gems/gems/vagrant-2.2.4/lib/vagrant/util/busy.rb:19:in `busy'
	from /opt/vagrant/embedded/gems/gems/vagrant-2.2.4/lib/vagrant/action/runner.rb:102:in `run'
	from /opt/vagrant/embedded/gems/gems/vagrant-2.2.4/lib/vagrant/machine.rb:238:in `action_raw'
	from /opt/vagrant/embedded/gems/gems/vagrant-2.2.4/lib/vagrant/machine.rb:209:in `block in action'
	from /opt/vagrant/embedded/gems/gems/vagrant-2.2.4/lib/vagrant/environment.rb:615:in `lock'
	from /opt/vagrant/embedded/gems/gems/vagrant-2.2.4/lib/vagrant/machine.rb:195:in `call'
	from /opt/vagrant/embedded/gems/gems/vagrant-2.2.4/lib/vagrant/machine.rb:195:in `action'
	from /opt/vagrant/embedded/gems/gems/vagrant-2.2.4/lib/vagrant/batch_action.rb:82:in `block (2 levels) in run'
```

### System configuration
**OS/Distro version:** Arch Linux: Kernel 5.0.3-arch1-1-ARCH
**Libvirt version:**
```
Compiled against library: libvirt 5.1.0
Using library: libvirt 5.1.0
Using API: QEMU 5.1.0
Running hypervisor: QEMU 3.1.0
```

**Output of `vagrant version; vagrant plugin list`:**
```
Installed Version: 2.2.4
Latest Version: 2.2.4

You're running an up-to-date version of Vagrant!
vagrant-libvirt (0.0.45, global)
vagrant-mutate (1.2.0, global)
vagrant-reload (0.0.1, global)
```

**Output of `VAGRANT_LOG=debug vagrant up --provider=libvirt`**
[vagrant.log](https://github.com/vagrant-libvirt/vagrant-libvirt/files/5647237/vagrant.log)

**A Vagrantfile to reproduce the issue:**
```
Vagrant.configure("2") do |config|
  config.vm.box = "archlinux/archlinux"

  config.vm.provider :libvirt do |domain|
    domain.memory = 4096
    domain.cpus = 4

    domain.qemu_use_session = true
    domain.system_uri = 'qemu:///system'
    domain.storage_pool_name = 'lfs'
    domain.storage :file,
                   :path => 'lfs_disk',
                   :device => 'vdb',
                   :size => '20G',
                   :allow_existing => true
  end
end
```
If I remove the `domain.storage :file ...` everything works fine.

**Are you using upstream vagrant package or your distros package?**
Distro, which is the same version as upstream, `2.2.4`

**Comment by mykaul** [comment]
I got the same stack. It worked with previous (0.4.4) plugin release (or vagrant - also upgraded to a newer vagrant release as part of upgrading my host host from Fedora 29 to Fedora 30).
I was not using QEMU directly, though.

My code:
```
          #disk_config
          if num != 0
            disks = []
            (1..(node_data_disk_count)).each do |d|
              lvt.storage :file, :device => "vd#{driveletters[d]}", :size => "#{disk_size}G"
              disks.push "/dev/vd#{driveletters[d]}"
            end #disks
```

**Comment by abalg** [comment]
I can confirm the same issue here ... using vagrant 2.2.3 / vagrant-libvirt-0.0.45 (Fedora 30) 

As soon as I add the second disk to the Vagrantfile vagrant crashes on "up"

**Comment by edoardo-c** [comment]
I get the same running 2.2.4 on Fedora 30.

More info and debug at: [Gist](https://gist.github.com/edoardo-c/4cf8d56fb400ceb0adaeb7da846e3210)

**Comment by clementmartin** [comment]
Same issue, unable to add a second disk.
Packages versions of fedora 30:
vagrant-2.2.4-2.fc30.noarch
vagrant-libvirt-0.0.45-1.fc30.noarch

**Comment by s2490** [comment]
Same issue on F30:
$ vagrant version
Installed Version: 2.2.5
$ vagrant plugin list
vagrant-libvirt (0.0.45, system)
vagrant-sshfs (1.3.1, system)

**Comment by guits** [comment]
same issue on F31
```bash
$ cat /etc/fedora-release 
Fedora release 31 (Thirty One)
$ vagrant --version
Vagrant 2.2.6
$ vagrant plugin list
vagrant-libvirt (0.0.45, system)
```

**Comment by cjsin** [comment]
I see this problem without qemu_use_session set to anything other than the default.

However I am using 
    lv.uri  = 'qemu:///system'

which I would have assumed is the same as disabling qemu_use_session.
However I guess not.

The following Vagrantfile is suitable to reprodue this with vagrant 2.2.4 
```
Vagrant.configure("2") do |config|
    config.vm.box = "centos/8"
    config.vm.provider "libvirt" do |lv, override|
        lv.storage :file, :device => "vdb", :path => "vagrant-bug-vdb.img", :size => "1G"
    end
end
```

In my case I have vagrant-libvirt-0.0.45, and it can be seen in:

    vagrant-libvirt-0.0.45/lib/vagrant-libvirt/action/create_domain.rb
    (line 136)

that now vagrant-libvirt only ever sets storage_prefix to any value at all, if config.qemu_use_session is not set.


It would seem from the code that adding disks is only possible at all when qemu_use_session is true.

I haven't used vagrant for perhaps a year or a bit more, but I have used it much in the past, with Vagrantfiles creating 12 VMs each with 3 or 4 disks, for testing Ceph clusters, and never had problems like this, just adding one data disk to a basic centos vm.

In my case, I found that if I remove the setting of :
    lv.uri = 'qemu:///system'
by:
   lv.qemu_use_session = false

it fixed the problem for me

How has this bug gone for a year without any attention, yet I'm able to find that it's the qemu_use_session variable (or an incompatible setting between this and the 'uri' setting), when the bug reporter mentioned it in the very first line of the bug report and it took me only 15 minutes to determine this.

Is this project abandoned?

**Comment by cjsin** [comment]
sorry for my poor language above, I meant to say:
In my case, I found that if I *remove* the setting of :
    lv.uri = 'qemu:///system'
and added, instead:
     lv.qemu_use_session = false
it fixed the problem (for me)

**Comment by cjsin** [comment]
the bug reporter mentioned that they had lv.qemu_use_session = true,  but it is just an idea that perhaps they had, like me, a different value of 'uri' which conflicted with their setting of qemu_use_session

**Comment by cjsin** [comment]
Either this is a bug, or else perhaps the documentation for 'libvirt.uri' and 'libvirt.qemu_use_session' needs to be updated, because it currently says:

"uri - For advanced usage. Directly specifies what Libvirt connection URI vagrant-libvirt should use. Overrides all other connection configuration options"

I would assume that qemu_use_session is one of the 'other connection configuration options'.

In 
/usr/share/vagrant/gems/gems/vagrant-libvirt-0.0.45/lib/vagrant-libvirt/config.rb:
function 'finalize
(approximately line 654)
this line:

@qemu_use_session = true if @qemu_use_session == UNSET_VALUE

should perhaps be updated to consider the @uri when @uri != UNSET_VALUE

(the code currently generates @uri if unset, just two lines below)

**Comment by cjsin** [comment]
I'm not a ruby dev but from looking at the code, my suggested workaround for users experiencing this problem, is, if using either 'uri' or the 'qemu_use_session' property, set them both to be consistent.
good luck

**Comment by electrofelix** [comment]
@cjsin I've recently merged #1062 via #1141 which included a fix to ensure set `qemu_use_session` if `uri` was to use a session, or it will correct construct `uri` if `qemu_use_session` is set. If the config specifies both by accident, an error will be output early instead of waiting until a problem occurs later.

However I've yet to check if the issue of creating storage using the session works because the part of the original issue here appears related to trying to use a system connection in the session mode will now be correctly flagged as incorrect, the next step is to make sure that once it's consistent with both settings the storage is created as expected in both cases.

**Comment by electrofelix** [comment]
Looks like we're missing another check to flag that setting `system_uri` and `qemu_use_session` to fail config validation. Currently the code will do so when using `uri` is set in a way that conflicts with `qemu_use_session`, but not if `system_uri` is set.

But otherwise, there is an issue at https://github.com/vagrant-libvirt/vagrant-libvirt/blob/master/lib/vagrant-libvirt/action/create_domain.rb#L167 due to https://github.com/vagrant-libvirt/vagrant-libvirt/blob/master/lib/vagrant-libvirt/action/create_domain.rb#L147-L156, so need to see if it's possible to resolve a storage_prefix if `qemu_use_session` is set.

After testing locally if I delete the check for whether qemu_use_session is set, and the pool is left at `default`, the code will create a suitable pool, if it's a custom pool that doesn't exist, it'll throw the expected error. So it looks like can just remove that check.

Unfortunately that blows up as it attempts to set the user/group to 0 on the resulting storage. It looks like the code just needs to be updated to follow the same pattern as https://github.com/vagrant-libvirt/vagrant-libvirt/blob/master/lib/vagrant-libvirt/action/create_domain_volume.rb#L45-L81 and provide the owner/group arguments using storage_uid/storage_gid and it worked locally, so should be possible to fix.
